lighthouse-core/gather/fetcher.js

```diff
@@ -0,0 +1,177 @@
+/**
+ * @license Copyright 2019 Google Inc. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+'use strict';
+
+/* global document */
+
+class Fetcher {
+  /**
+   * @param {import('./driver.js')} driver
+   */
+  constructor(driver) {
+    this.driver = driver;
+    /** @type {Map<string, (event: LH.Crdp.Fetch.RequestPausedEvent) => void>} */
+    this._onRequestPausedHandlers = new Map();
+    this._onRequestPaused = this._onRequestPaused.bind(this);
+  }
+
+  /**
+   * The Fetch domain accepts patterns for controlling what requests are intercepted, but we
+   * enable the domain for all patterns and filter events at a lower level to support multiple
+   * concurrent usages. Reasons for this:
+   *
+   * 1) only one set of patterns may be applied for the entire domain.
+   * 2) every request that matches the patterns are paused and only resumes when certain Fetch
+   *    commands are sent. So a listener of the `Fetch.requestPaused` event must either handle
+   *    the requests it cares about, or explicitly allow them to continue.
+   * 3) if multiple commands to continue the same request are sent, protocol errors occur.
+   *
+   * So instead we have one global `Fetch.enable` / `Fetch.requestPaused` pair, and allow specific
+   * urls to be intercepted via `driver.setOnRequestPausedHandler`.
+   */
+  async enableRequestInterception() {
+    await this.driver.sendCommand('Fetch.enable', {
+      patterns: [{requestStage: 'Request'}, {requestStage: 'Response'}],
+    });
+    await this.driver.on('Fetch.requestPaused', this._onRequestPaused);
+  }
+
+  /**
+   * @param {string} url
+   * @param {(event: LH.Crdp.Fetch.RequestPausedEvent) => void} handler
+   */
+  async setOnRequestPausedHandler(url, handler) {
```

Comment:
should this be private? do we expect someone external to use this and/or would it interfere with the operation of `fetchResource` if they used it incorrectly?

Reply:
yeah it should be private

Reply:
at least until there is a use case for someone to actually use the fetch interception for nontrivial stuff :)

Reply:
actually, not convinced it should be private. this whole class is meant to make the Fetch domain available to LH. maybe Fetcher is the wrong name?

Reply:
> this whole class is meant to make the Fetch domain available to LH. maybe Fetcher is the wrong name?

:/ do we need to do that? AFAIK by far our most pressing usage of this will be to fetch things we currently do through `evaluateAsync` (and the only usage in this PR, right?)

Reply:
ok:)

Reply:
so...... private? :)

Reply:
yes i forgot.

recently OH: all good things are worth waiting for